### DISCUSSION TITLE
data structure having posts, comments, and users

### DISCUSSION THREAD

**Discussion by niklasrammelkamp** [neuefische/questions#483] [Web] [answered]
Hi there,
in my latest pull request, I structured my data completely new. 
Before, I had a single array containing all posts and their respective comments. 

Now I have a posts, comments, and a users array. 
Each post has a reference to its comments, what user uploaded it, and who left a like. 
Each comment has a reference to what post it belongs to and who wrote the comment.
And each user has a reference to what posts were uploaded, what posts were liked, and the written comments.

If you are familiar with similar cases and already know what problems I can get with this, please let me know if you have any suggestions for improvement.

Here is the [pull request](https://github.com/niklasrammelkamp/capstone-project/pull/22). 

Thank you 🫶

**Comment by ThomasF85** [comment, marked as answer]
Hi Niklas,

your question is not one where there is one definite best way of approaching designing the data structure. It all depends on your queries, how many relations there are between your data structures, what your typical use cases of CRUD operations on your data structures are.
There are a lot of things to consider, most notably:
1. The need for and how to achieve atomic updates (we don't want to delete a post and then have the server crash before it can delete the comments for that post. That would leave us with orphaned references and potentially the app crashing.)
2. Efficiency gains vs. data duplication

Here is an article with some rules of thumb: https://www.mongodb.com/blog/post/6-rules-of-thumb-for-mongodb-schema-design

The main take aways are:
- A one to few relationship is a good use case for embedding instead of references
- A one to many relationship might be a good use case for one side referencing where the references lie on the "one" side
- A one to super many relationship might be a good use case for one side referencing where the references lie on the "many" side
- Two way referencing makes reading easier but writing more complicated

**Reply by niklasrammelkamp** [reply]
Thank you. The article is pretty helpful. Thank you :)
